File: services/admin/src/routes/audit.py
"""Audit log routes."""

from fastapi import APIRouter, Depends, HTTPException, Query

from src.middleware.auth import get_current_user
from src.models.responses import AuditLogsResponse, AuditLogResponse
from src.services.audit.get_audit_logs import get_audit_logs

router = APIRouter(prefix="/audit-logs", tags=["admin-audit"])


@router.get("", response_model=AuditLogsResponse)
async def get_audit_logs_endpoint(
    page: int = Query(1, ge=1, description="Page number (1-indexed)"),
    limit: int = Query(12, ge=1, le=100, description="Records per page"),
    action: str = Query(None, description="Filter by action"),
    entity_type: str = Query(None, description="Filter by entity type"),
    user_role: str = Query(None, description="Filter by user role"),
    search: str = Query(None, description="Search by user name or entity reference"),
    _: dict = Depends(get_current_user),
):
    """Get audit logs with pagination and optional filters."""
    try:
        return await get_audit_logs(page, limit, action, entity_type, user_role, search)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception:
        raise HTTPException(status_code=500, detail="Failed to retrieve audit logs")

File: apps/webapp/src/services/admin/services/audit.ts
import { apiClient } from '@/services/apiClient';
import type { AuditLog, AuditLogFilters } from '@/types/audit';

export interface PaginatedAuditLogs {
  logs: AuditLog[];
  total: number;
  page: number;
  pages: number;
  limit: number;
}

export interface GetAuditLogsParams extends AuditLogFilters {
  page?: number;
  limit?: number;
}

/**
 * Get paginated audit logs with optional filters
 */
export async function getAuditLogs(params?: GetAuditLogsParams): Promise<PaginatedAuditLogs> {
  const searchParams = new URLSearchParams();

  if (params) {
    if (params.page) searchParams.append('page', params.page.toString());
    if (params.limit) searchParams.append('limit', params.limit.toString());
    if (params.user_id) searchParams.append('user_id', params.user_id);
    if (params.action) searchParams.append('action', params.action);
    if (params.entity_type) searchParams.append('entity_type', params.entity_type);
    if (params.user_role) searchParams.append('user_role', params.user_role);
    if (params.entity_id) searchParams.append('entity_id', params.entity_id);
    if (params.start_date) searchParams.append('start_date', params.start_date);
    if (params.end_date) searchParams.append('end_date', params.end_date);
    if (params.search) searchParams.append('search', params.search);
  }

  const queryString = searchParams.toString();
  const url = queryString ? `/admin/audit-logs?${queryString}` : '/admin/audit-logs';

  const response = await apiClient.get<PaginatedAuditLogs>(url);
  return response;
}
